File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/design_linked_list.py
class MyNode(object):
    
    def __init__(self, val):
        self.val = val
        self.next = None


class MyLinkedList(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.length = 0
        self.head = None
        

    def get(self, index):
        """
        Get the value of the index-th node in the linked list. If the index is invalid, return -1.
        :type index: int
        :rtype: int
        """
        if index < 0 or index >= self.length:
            return -1
        if index == 0:
            return self.head.val
        else:
            current = self.head
            for i in range(index):
                current = current.next
            return current.val
        

    def addAtHead(self, val):
        """
        Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
        :type val: int
        :rtype: None
        """
        first = MyNode(val)
        first.next = self.head
        self.head = first
        self.length += 1
        

    def addAtTail(self, val):
        """
        Append a node of value val to the last element of the linked list.
        :type val: int
        :rtype: None
        """
        last = MyNode(val)
        current = self.head
        if self.length == 0:
            self.head = last
            self.length += 1
        else:
            for _ in range(self.length-1):
                current = current.next
            current.next = last
            self.length += 1
        

    def addAtIndex(self, index, val):
        """
        Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
        :type index: int
        :type val: int
        :rtype: None
        """
        if index < 0 or index > self.length:
            return;
        elif index == 0:
            self.addAtHead(val)
        elif index == self.length:
            self.addAtTail(val)
        else:
            insert = MyNode(val)
            prev = self.head
            for _ in range(index-1):
                prev = prev.next
            succ = prev.next
            prev.next = insert
            insert.next = succ
            self.length += 1
        

    def deleteAtIndex(self, index):
        """
        Delete the index-th node in the linked list, if the index is valid.
        :type index: int
        :rtype: None
        """
        if index < 0 or index >= self.length:
            return
        elif index == 0:
            self.head = self.head.next
            self.length -= 1
        else:
            current = self.head
            for _ in range(index-1):
                current = current.next
            current.next = current.next.next
            self.length -= 1


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/DesignLinkedList.cs
public class MyNode {
    public int val;
    public MyNode next;
    public MyNode(int val) {
        this.val = val;
    }
}


public class MyLinkedList {
    public int length;
    public MyNode head;

    /** Initialize your data structure here. */
    public MyLinkedList() {
        length = 0;
        head = null;
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    public int Get(int index) {
        if (index < 0 || index >= length) return -1;
        if (index == 0) return head.val;
        else {
            MyNode current = head;
            for (int i=0; i<index; i++) {
                current = current.next;
            }
            return current.val;
        }
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    public void AddAtHead(int val) {
        MyNode first = new MyNode(val);
        first.next = head;
        head = first;
        length++;
    }
    
    /** Append a node of value val to the last element of the linked list. */
    public void AddAtTail(int val) {
        MyNode last = new MyNode(val);
        MyNode current = head;
        if (length == 0) {
            head = last;
            length++;
        } else {
            for (int i=0; i<length-1; i++) {
                current = current.next;
            }
            current.next = last;
            length++;
        }
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    public void AddAtIndex(int index, int val) {
        if (index < 0 || index > length) AddAtTail(val);
        else if (index == 0) AddAtHead(val);
        else {
            MyNode insert = new MyNode(val);
            MyNode prev = head;
            MyNode next;
            for (int i=0; i<index-1; i++) {
                prev = prev.next;
            }
            next = prev.next;
            prev.next = insert;
            insert.next = next;
            length++;
        }
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    public void DeleteAtIndex(int index) {
        if (index < 0 || index >= length) return;
        else if (index == 0) {
            head = head.next;
            length--;
        } else {
            MyNode current = head;
            for (int i=0; i<index-1; i++) {
                current = current.next;
            }
            current.next = current.next.next;
            length--;
        }
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */
